### PR TITLE
Support undeleting keys

### PR DESCRIPTION
**Pull request by lmars**
Due to the fact that deleting a key doesn't actually delete it from the db (it just sets it's deleted_at time), re-adding a key which has been deleted currently causes a 500 error with a log message of:

```
duplicate key value violates unique constraint "keys_pkey"
```

I have updated the logic so that if the key being added is a deleted key, set it's deleted_at back to NULL.

### CODE REVIEW

**Comment by titanous** [comment]
Hmm, instead of undeleting we should probably change to a UUID primary key and add a unique index on the fingerprint with a `WHERE deleted_at IS NULL` condition.

**Comment by lmars** [comment]
@titanous yes the change to a UUID primary key makes sense, see 7c83cba.

I have assumed at this early stage that there is no need for a new migration and have just changed the initial migration step.

Also, this change breaks flynn-cli (e.g. [here](https://github.com/flynn/flynn-cli/blob/master/keys.go#L44)) so I'm not sure how we want to handle that? Just change it and expect people to be using the latest controller?

**Comment by titanous** [comment]
No need to break anything, keep using the fingerprint as the ID in requests/responses.

**Comment by lmars** [comment]
Yep, sounds sensible: 9761f9f

File: key.go [hunk 1]
```diff
@@ -53,16 +53,16 @@ func scanKey(s Scanner) (*ct.Key, error) {
...
 
 func (r *KeyRepo) Get(id string) (interface{}, error) {
-	row := r.db.QueryRow("SELECT key_id, key, comment, created_at FROM keys WHERE key_id = $1 AND deleted_at IS NULL", id)
+	row := r.db.QueryRow("SELECT fingerprint, key, comment, created_at FROM keys WHERE fingerprint = $1 AND deleted_at IS NULL", id)
 	return scanKey(row)
 }
 
 func (r *KeyRepo) Remove(id string) error {
-	return r.db.Exec("UPDATE keys SET deleted_at = now() WHERE key_id = $1", id)
+	return r.db.Exec("UPDATE keys SET deleted_at = now() WHERE fingerprint = $1", id)
```
Comment (titanous) [inline key.go]: Needs a `AND deleted_at IS NULL`.

Comment (lmars) [inline key.go]: Sure: db2bdae

**Comment by progrium** [comment]
Why do we keep around deleted keys?

On Fri, Apr 25, 2014 at 10:05 AM, Jonathan Rudenberg <
notifications@github.com> wrote:

> Merged #6 https://github.com/flynn/flynn-controller/pull/6.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/flynn/flynn-controller/pull/6
> .

## 

Jeff Lindsay
http://progrium.com

**Comment by titanous** [comment]
@progrium Everything is soft deleted, just a habit (I try to never call `DELETE`). Any reason not to keep them around?

**Comment by progrium** [comment]
Not, just thought there might have been something else to it.

On Mon, Apr 28, 2014 at 6:32 PM, Jonathan Rudenberg <
notifications@github.com> wrote:

> @progrium https://github.com/progrium Everything is soft deleted, just
> a habit (I try to never call DELETE). Any reason not to keep them around?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/flynn/flynn-controller/pull/6#issuecomment-41627070
> .

## 

Jeff Lindsay
http://progrium.com

**Pull request merged**